Sources/Authgear.swift

```diff
@@ -882,35 +881,34 @@ public class Authgear: NSObject {
                     // registerCurrentWeChatRedirectURI will be called and overwrite
                     // previous registered wechatRedirectURI
                     self.registerCurrentWechatRedirectURI(uri: wechatRedirectURI)
-
-                    let vc = UIViewController()
-                    let wv = WKWebView(frame: vc.view.bounds)
-                    wv.autoresizingMask = [.flexibleWidth, .flexibleHeight]
-                    wv.navigationDelegate = self
-                    wv.load(URLRequest(url: endpoint))
-                    vc.view.addSubview(wv)
-                    vc.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(self.dismissWebView))
-                    self.webViewViewController = vc
-
-                    let nav = UINavigationController(rootViewController: vc)
-                    nav.modalPresentationStyle = .pageSheet
-
-                    let window = UIApplication.shared.windows.filter { $0.isKeyWindow }.first
-                    window?.rootViewController?.present(nav, animated: true) {
-                        handler?(.success(()))
-                    }
+                    self.authenticationSession = self.authenticationSessionProvider.makeAuthenticationSession(
+                        url: endpoint,
+                        // Opening an arbitrary does not have a clear goal.
+                        // So here we pass a placeholder callbackURL scheme.
+                        callbackURLSchema: "nocallback",
+                        // prefersEphemeralWebBrowserSession is true so that
+                        // the alert dialog is never prompted and
+                        // the app session token cookie is forgotten when the webview is closed.
+                        prefersEphemeralWebBrowserSession: true,
+                        completionHandler: { [weak self] result in
+                            self?.unregisterCurrentWechatRedirectURI()
+                            switch result {
+                            case .success:
+                                // This branch is unreachable.
+                                handler?(.success(()))
+                            case let .failure(error):
+                                handler?(.failure(error))
+                            }
+                        }
+                    )
+                    self.authenticationSession?.start()
```

Comment:
For porting the original webview behavior, should we call `handler?(.success(()))` right after the window is opened? And in the ASWebAuthenticationSession completion handler, we should only do some cleanup job and `handler?(.failure(error))` should not be called.

Reply:
After our offline discussion, we are going to change the behavior of openURL. The callback will be called when the webview is about to close. When the user cancels the webview, it is considered as success instead of cancel.